### PR TITLE
Enhancement/metric terms

### PR DESCRIPTION
**Pull request by hgd511**
Calculates the local Grad-Shafranov solution using local geometry, which is used to calculate the covariant and contravariant metric tensor components for the toroidal system and the field-aligned system.

The additional derivatives in R and Z necessary for these computations have been included in the relevant LocalGeometry files. 

Because the equivalent of ff_prime can be calculated in this formalism independently of its measured value, a comparison between and the measurement and the calculation of ff_prime can provide a measure for the quality of the geometric parameterisation.

Attached is a file containing basic documentation of the different quantities that can be calculated using MetricTerms. Please advise regarding the best place for this.
[PyroMetricTerms.pdf](https://github.com/pyro-kinetics/pyrokinetics/files/10897087/PyroMetricTerms.pdf)

### CODE REVIEW

File: pyrokinetics/local_geometry/LocalGeometry.py [hunk 1]
```diff
@@ -525,19 +529,21 @@ def plot_equilibrium_to_local_geometry_fit(
...
         -------
         Copy of LocalGeometry object
-        """
+        ""
         # Create new empty object. Works for derived classes too.
         new_localgeometry = self.__class__()
         for key, value in self.items():
             new_localgeometry[key] = deepcopy(value, memodict)
         return new_localgeometry
+    """
```
Comment (bpatel2107) [inline pyrokinetics/local_geometry/LocalGeometry.py]: Can probably delete this section

**Comment by bpatel2107** [comment]
Overall this looks good!

- We should add some tests for some basic circular plasmas where the metric terms have a simple geometric definitions.
- @ZedThree @LiamPattinson should we put that PDF in docs or is there a nice way to embed in into the readthedocs?
- The tests are currently failing due to the different definitions of "Miller" that codes have (see #143) . When only including elongation and triangularity, the codes are all the same but when including squareness, the codes diverge so we should find a way to handle this. The first thing I can thing of is to have a `BasicMiller` geometry that doesn't have squareness that we can use for cross-code comparisons/roundtrips, but then have the option to use squareness in something like `FullMiller` geometry

**Comment by ZedThree** [comment]
Well, seeing as @hgd511 wrote the PDF, we can just convert the LaTeX to reStructuredText and include it straight in the docs, keeping all the equations intact :slightly_smiling_face: 

@hgd511 The easiest thing perhaps is to use [pandoc](https://pandoc.org) to convert your source file to rst, and then add it to [docs/rst_docs/local_geometry.rst](https://github.com/pyro-kinetics/pyrokinetics/blob/unstable/docs/rst_docs/local_geometry.rst)

**Comment by bpatel2107** [comment]
@ZedThree @LiamPattinson This PR looks good but the failing tests requires quite a big refactor to fix. Is it worth merging this into unstable and temporarily not run those tests, or open a separate PR to fix that, merge that into this and then merge everything into unstable

**Comment by ZedThree** [comment]
I would definitely get the tests fixed before merging!

**Comment by bpatel2107** [comment]
Big ol refactor coming in then!

**Comment by LiamPattinson** [comment]
As easy(-ish) fix might be to catch those `NotImplementedError`s in a `try...except` block, and if the provided `LocalGeometry` type doesn't match the target GK code, it should raise a warning and instead fit a compatible `LocalGeometry`. So rather than saying "GS2 doesn't support MXH" and failing, it could instead just fit a `LocalGeometryMiller` as best as it can.

**Review by bpatel2107 — APPROVED** [review]
LGTM

**Pull request merged**